### DISCUSSION TITLE
Anyway to record videos for browsers sessions launched via "driver = Driver()" class?

### DISCUSSION THREAD

**Discussion by iamumairayub** [seleniumbase/SeleniumBase#2370] [Q&A] [answered]
I am on Ubuntu VPS and following piece of code works perfectly.
I run my code with `python myfile.py` command.
What I want is to record and save video on my same server.

```
with Display(visible=0, size=(1440, 1880)):

    driver = Driver(
        uc=True, 
        browser='chrome',
        headless2=True,
    )

    driver.get('https://www.whatismybrowser.com/')
    driver.quit()
```

**Comment by mdmintz** [comment, marked as answer]
For video recording, see: https://github.com/seleniumbase/SeleniumBase/issues/642#issuecomment-666353300:

Use an existing tool with SeleniumBase such as Zalenium for recording videos:
https://opensource.zalando.com/zalenium/
It works like a Selenium Grid.
See https://seleniumbase.io/seleniumbase/utilities/selenium_grid/ReadMe/ for instructions on using SeleniumBase to connect to a Selenium Grid.

However, that requires the use of a SeleniumBase format (`SB()` or `BaseCase`) since the `Driver()` format doesn't have the necessary code. Also, UC Mode isn't supported for Selenium Grid because SeleniumBase needs direct access to a local driver for that to work. As mentioned in https://github.com/seleniumbase/SeleniumBase/issues/2278#issuecomment-1810506869:

UC Mode doesn't support remote browsers. The driver and browser must both be on a local machine so that UC Mode can make modifications to the driver (transforming `chromedriver` into `uc_driver`), as well as being able to spin up Chrome from the backend so that it can be directly connected, disconnected, and reconnected again as needed.

**Reply by iamumairayub** [reply]
I read your reply and checked above links and found out that its only possible to record video for remote browsers?

<hr>

Lets forget about `Driver()` and **Remote Browser**s for a moment. 

Let say we have this code

```
with SB(uc=True) as sb:
    sb.open("https://www.amazon.co.uk/s?k=test&crid=16478HWXKS706&sprefix=%2Caps%2C571&ref=nb_sb_noss")
    sb.slow_scroll_to_element('a.s-pagination-item.s-pagination-next', by="css selector")
```

`uc=True` or `uc=False`, is it possible to record this session?

there is `activate_recorder()` but its not for video recording.

**Reply by mdmintz** [reply]
You can record the session by spinning up a Selenium Grid first. There are two options:
* (https://opensource.zalando.com/zalenium/ ---- https://github.com/zalando/zalenium)
* https://github.com/SeleniumHQ/docker-selenium?tab=readme-ov-file#video-recording

Each of those has instructions for video recording.

To connect to a grid with seleniumbase, set the `server` and `port` parameters. Eg:

```python
with SB(server="127.0.0.1", port="4444") as sb:
    sb.open("URL")
    # Do more things
```

Follow the instructions from the external sites for configuring a Selenium Grid for recording video.

(Note that Selenium Grid Mode doesn't support UC Mode.)

**Reply by iamumairayub** [reply]
We are not using remote selenium. We are using local selenium with `Display` on our Ubuntu VPS.
Is it possible to record videos of locally running sessions? (not remote sessions)

**Reply by mdmintz** [reply]
Selenium Grid can be run locally by setting the server to `"127.0.0.1"`.
It's the only way I know to automatically record videos during test runs.

**Reply by iamumairayub** [reply]
Thanks for your replies Minz.

But we dont want to run SeleniumGrid because that does not let us use undetected_chromedriver.

Regarding 

> Selenium Grid can be run locally by setting the server to "127.0.0.1".

I would like to bring this to you that Selenoid enables you to record videos of your Remote sessions by just passing `enableVideo: true` as option

`options.set_capability("selenoid:options", {"enableVideo": True})`

https://aerokube.com/selenoid/latest/#_video_recording_enablevideo_videoname_videoscreensize_videoframerate_videocodec

I dont expect to your further replies on this thread. Thanks again Minz.
